ndcube/ndcube.py

```diff
@@ -952,3 +962,234 @@ def to(self, new_unit, **kwargs):
+                 # As a silly example, propagate uncertainties by summing those in same bin.
+                 # Note not all args are used, but function must accept them.
+                 n_pixels_per_bin = data.shape[0]  # 1st dimension of inputs gives pixels in bin.
+                 final_shape = data.shape[1:]  # Trailing dims give shape of put rebinned data.
+                 # Propagate uncerts by adding them.
+                 new_uncert = numpy.zeros(final_shape)
+                 for i in range(n_pixels_per_bin):
+                     new_uncert += uncertainty.array[i]
+                 # Alternatively: new_uncerts = uncertainty.array.sum(axis=0)
+                 return type(uncertainty)(new_uncert)  # Convert to original uncert type and return.
+        """
+        # Sanitize input.
+        if new_unit is None:
+            new_unit = self.unit
+        # Make sure the input bin dimensions are integers.
+        bin_shape = np.rint(bin_shape).astype(int)
+        offsets = (bin_shape - 1) / 2
+        if all(bin_shape == 1):
+            return self
+        # Ensure bin_size has right number of entries and each entry is an
+        # integer fraction of the array shape in each dimension.
+        data_shape = self.dimensions.value.astype(int)
+        naxes = len(data_shape)
+        if len(bin_shape) != naxes:
+            raise ValueError("bin_shape must have an entry for each array axis.")
+        if (np.mod(data_shape, bin_shape) != 0).any():
+            raise ValueError(
+                "bin shape must be an integer fraction of the data shape in each dimension. "
+                f"data shape: {data_shape};  bin shape: {bin_shape}")
+
+        # Reshape array so odd dimensions represent pixels to be binned
+        # then apply function over those axes.
+        m = None if (self.mask is None or self.mask is False or use_masked_values) else self.mask
+        data = self.data
+        if m is None:
+            data = self.data
+        else:
+            masked_type = np.ma.masked_array
+            for array_type, m_type in ARRAY_MASK_MAP.items():
+                if isinstance(self.data, array_type):
+                    masked_type = m_type
+                    break
+            else:
+                warn.warning("data and mask arrays of different or unrecognized types. "
+                             "Casting them into a numpy masked array.")
+            data = masked_type(self.data, m)
+        reshape = np.empty(data_shape.size + bin_shape.size, dtype=int)
+        new_shape = (data_shape / bin_shape).astype(int)
+        reshape[0::2] = new_shape
+        reshape[1::2] = bin_shape
+        reshape = tuple(reshape)
+        reshaped_data = data.reshape(reshape)
+        operation_axes = tuple(range(len(reshape) - 1, 0, -2))
+        new_data = operation(reshaped_data, axis=operation_axes)
+        if isinstance(new_data, ARRAY_MASK_MAP[np.ndarray]):
+            new_data = new_data.data
```

Comment:
This still weirds me out. It feels like we have special cased dask and numpy here, where numpy does this but dask dosen't. What if cupy adds a masked array where we need to do this. It all makes me uncomfortable.

Reply:
You're right.  But since numpy and dask masked arrays have different APIs, the alternative I see is to force everything into numpy masked arrays which is a worse idea.